src/python/pants/option/compiler_option_sets_mixin.py

```diff
@@ -0,0 +1,99 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from builtins import object
+
+
+class CompilerOptionSetsMixin(object):
+  """A mixin for language-scoped that support compiler option sets."""
+
+  @classmethod
+  def register_options(cls, register):
+    super(CompilerOptionSetsMixin, cls).register_options(register)
+
+    register('--fatal-warnings-enabled-args', advanced=True, type=list, fingerprint=True,
+             default=list(cls.get_fatal_warnings_enabled_args_default()),
+             help='Extra compiler args to use when fatal warnings are enabled.')
+    register('--fatal-warnings-disabled-args', advanced=True, type=list, fingerprint=True,
+             default=list(cls.get_fatal_warnings_disabled_args_default()),
+             help='Extra compiler args to use when fatal warnings are disabled.')
+    register('--compiler-option-sets-enabled-args', advanced=True, type=dict, fingerprint=True,
+             default=cls.get_compiler_option_sets_enabled_default_value(),
+             help='Extra compiler args to use for each enabled option set.')
+    register('--compiler-option-sets-disabled-args', advanced=True, type=dict, fingerprint=True,
```

Comment:
Is there a reason we can't make these `default=` methods into class fields that subclasses can override (and they can use `@classproperty` from `pants.util.meta` for dynamic stuff) instead of methods? That seems simpler with no obvious downside to me.

Reply:
Subclasses are overriding this method in their respective code currently as classmethods. I'm not opposed, but how does it simplify this?